### PR TITLE
fixes #2929 - generate encryption key and encrypt data in postinstall

### PR DESCRIPTION
**Pull request by domcleal**
Tested clean installs and upgrades on EL6 successfully.

Debian version: https://github.com/theforeman/foreman-packaging/pull/191

### CODE REVIEW

File: foreman.spec [hunk 1]
```diff
@@ -504,7 +505,13 @@ rm -rf %{buildroot}
...
 %attr(-,%{name},root) %{_datadir}/%{name}/config.ru
 %attr(-,%{name},root) %{_datadir}/%{name}/config/environment.rb
-%ghost %attr(0640,root,%{name}) %{_datadir}/%{name}/config/initializers/local_secret_token.rb
+# Symlink to /etc, EL6 needs attrs for ghost files, Fedora doesn't
+%if 0%{?rhel} == 6
+%ghost %attr(0777,root,root) %{_datadir}/%{name}/config/initializers/encryption_key.rb
+%else
+%ghost %{_datadir}/%{name}/config/initializers/encryption_key.rb
+%endif
+%ghost %attr(0640,root,%{name}) %config(noreplace) %{_datadir}/%{name}/config/initializers/local_secret_token.rb
```
Comment (lzap) [inline foreman.spec:514]: Why does RHEL6 need extra attrs while on the second line you keep it for both Fedora and RHEL6? I miss the reason here.

Comment (domcleal) [inline foreman.spec:514]: You had to ask... :)

On EL6, %attr is required for every %ghost file.

On Fedora 19, %attr is optional for %ghost files.  However....
- if the file is marked as %config, it doesn't have to exist during rpmbuild (after all, it's a ghost)
- if it's marked as a %dir, it doesn't have to exist ([fixed in F18](https://bugzilla.redhat.com/show_bug.cgi?id=839656))
- if it's marked only as %ghost and has 0777 attributes (it's a symlink), then rpm seems to assume it's a directory, but strangely requires that it's present in the buildroot (seems this is a variant of the above BZ, which I hit while testing this)

So it seems you need to leave out attrs to get it to build on F19 (else it fails trying to find the ghost), but you need them on EL6 all the time.

**Comment by lzap** [comment]
Hmmm. Works for me, but I do not like spawning rails in the build section, this is how Katello generates those:

```
#Generate secret token if the file does not exist
#(this must be called both for installation and upgrade)
TOKEN=/etc/katello/secret_token
# this file must not be world readable at generation time
umask 0077
test -f $TOKEN || (echo $(</dev/urandom tr -dc A-Za-z0-9 | head -c128) > $TOKEN \
    && chmod 600 $TOKEN && chown katello:katello $TOKEN)
```

But if you insist, I can live with that :-)

**Comment by domcleal** [comment]
Well we launch it anyway for db:migrate, so this additional task has no startup overhead (it's part of the same command).  And @isratrade has already done the work of printing out a nicely formatted file with comments etc :)

**Comment by domcleal** [comment]
@lzap?  The Debian half has been merged.

**Comment by lzap** [comment]
The moment you provide me with a build, I am starting with the review :-D

http://koji.katello.org/koji/taskinfo?taskID=92878

**Comment by domcleal** [comment]
Looks like it got garbage collected.  [test]

**Comment by domcleal** [comment]
Random [test] failure, but here are the EL6 RPMs: http://koji.katello.org/koji/taskinfo?taskID=92908

File: foreman.spec [hunk 2]
```diff
@@ -551,19 +558,39 @@ if [ ! -d $varlibdir/%{name} -a -d $datadir/tmp -a ! -L $datadir/tmp ]; then
...
+# secret token used for cookie signing etc.
 if [ ! -f %{_datadir}/%{name}/config/initializers/local_secret_token.rb ]; then
   touch %{_datadir}/%{name}/config/initializers/local_secret_token.rb
-  chmod 0640 %{_datadir}/%{name}/config/initializers/local_secret_token.rb
+  chmod 0660 %{_datadir}/%{name}/config/initializers/local_secret_token.rb
   chgrp foreman %{_datadir}/%{name}/config/initializers/local_secret_token.rb
-  %{scl_rake} -f %{_datadir}/%{name}/Rakefile security:generate_token >/dev/null 2>&1 || :
+  %{foreman_rake} security:generate_token >/dev/null 2>&1 || :
+  chmod 0640 %{_datadir}/%{name}/config/initializers/local_secret_token.rb
+fi
```
Comment (lzap) [inline foreman.spec:568]: Note - the local_secret_token.rb is not writeable to foreman user, so foreman-rake could not be used anyway. But if it exists, the code above is not executed anyway.

**Comment by lzap** [comment]
Upgrades work fine, I am gonna test clean install.

**Comment by lzap** [comment]
```
[lzap@lzapx SetupKojiRepo]$ mybkr -t /Satellite-6/Installation/SetupKojiRepo -t /Satellite-6/Installation/Foreman-develop -p KOJI_BUILD='http://koji.katello.org/koji/taskinfo?taskID=92908'
Submitting job MYBKR RHEL-6.5  
Submitted: ['J:618778']
```

https://github.com/lzap/bin-public/blob/master/mybkr

:coffee: :cake: :dancers:

**Comment by domcleal** [comment]
@lzap did it work? ;)

**Comment by lzap** [comment]
Nei, spawned another one :-( 'J:623600' Error in my test script.

**Comment by lzap** [comment]
My super-testing beaker script does not work well, working on this.

**Comment by lzap** [comment]
Ok fixed tests, hopefully soon, sorry about the delay.

**Comment by lzap** [comment]
Omg I am tuning my tests, even rewritten my testing beaker script into bats and the root cause is - the scratch build expired, RPMs were deleted :-) Rebuilding: http://koji.katello.org/koji/taskinfo?taskID=97091

**Comment by lzap** [comment]
Ok finally - I cannot confirm this to work. This is what I get:

```
==> /var/log/foreman/db_migrate.log <==
rake aborted!
Please install the sqlite3 adapter: `gem install activerecord-sqlite3-adapter` (Could not find sqlite3 (~> 1.3.5) amongst [actionmailer-3.2.8, actionpack-3.2.8, activemodel-3.2.8, activerecord-3.2.8, activeresource-3.2.8, activesupport-3.2.8, ancestry-2.0.0, apipie-rails-0.1.2, arel-3.0.2, audited-3.0.0, audited-activerecord-3.0.0, bigdecimal-1.1.0, bootstrap-sass-3.0.3.0, builder-3.0.0, bundler-1.1.4, bundler_ext-0.3.0, daemons-1.1.4, deep_cloneable-1.6.0, diff-lcs-1.1.3, erubis-2.7.0, eventmachine-0.12.10, fast_gettext-0.8.0, foreigner-1.4.2, gettext_i18n_rails-0.10.0, gettext_i18n_rails_js-0.0.8, hike-1.2.1, i18n-0.6.0, i18n_data-0.2.7, io-console-0.3, journey-1.0.4, jquery-rails-2.0.2, jquery-ui-rails-4.0.2, json-1.5.5, mail-2.4.4, mime-types-1.19, multi_json-1.8.2, net-http-persistent-2.7, net-ldap-0.3.1, oauth-0.4.7, pg-0.12.2, po_to_json-0.0.7, polyglot-0.3.3, rabl-0.9.0, rack-1.4.1, rack-cache-1.2, rack-ssl-1.3.2, rack-test-0.6.1, rails-3.2.8, railties-3.2.8, rake-0.9.2.2, rdoc-3.9.5, ref-1.0.0, rest-client-1.6.1, ruby2ruby-2.0.1, ruby_parser-3.1.1, safemode-1.2.0, sass-3.2.13, scoped_search-2.7.1, sexp_processor-4.1.3, sprockets-2.4.5, therubyracer-0.11.0beta5, thin-1.3.1, thor-0.15.4, tilt-1.3.3, treetop-1.4.10, tzinfo-0.3.33, uuidtools-2.1.3, will_paginate-3.0.2])

Tasks: TOP => db:migrate => environment
(See full trace by running task with --trace)

==> /var/log/foreman/db_seed.log <==
rake aborted!
Please install the sqlite3 adapter: `gem install activerecord-sqlite3-adapter` (Could not find sqlite3 (~> 1.3.5) amongst [actionmailer-3.2.8, actionpack-3.2.8, activemodel-3.2.8, activerecord-3.2.8, activeresource-3.2.8, activesupport-3.2.8, ancestry-2.0.0, apipie-rails-0.1.2, arel-3.0.2, audited-3.0.0, audited-activerecord-3.0.0, bigdecimal-1.1.0, bootstrap-sass-3.0.3.0, builder-3.0.0, bundler-1.1.4, bundler_ext-0.3.0, daemons-1.1.4, deep_cloneable-1.6.0, diff-lcs-1.1.3, erubis-2.7.0, eventmachine-0.12.10, fast_gettext-0.8.0, foreigner-1.4.2, gettext_i18n_rails-0.10.0, gettext_i18n_rails_js-0.0.8, hike-1.2.1, i18n-0.6.0, i18n_data-0.2.7, io-console-0.3, journey-1.0.4, jquery-rails-2.0.2, jquery-ui-rails-4.0.2, json-1.5.5, mail-2.4.4, mime-types-1.19, multi_json-1.8.2, net-http-persistent-2.7, net-ldap-0.3.1, oauth-0.4.7, pg-0.12.2, po_to_json-0.0.7, polyglot-0.3.3, rabl-0.9.0, rack-1.4.1, rack-cache-1.2, rack-ssl-1.3.2, rack-test-0.6.1, rails-3.2.8, railties-3.2.8, rake-0.9.2.2, rdoc-3.9.5, ref-1.0.0, rest-client-1.6.1, ruby2ruby-2.0.1, ruby_parser-3.1.1, safemode-1.2.0, sass-3.2.13, scoped_search-2.7.1, sexp_processor-4.1.3, sprockets-2.4.5, therubyracer-0.11.0beta5, thin-1.3.1, thor-0.15.4, tilt-1.3.3, treetop-1.4.10, tzinfo-0.3.33, uuidtools-2.1.3, will_paginate-3.0.2])

Tasks: TOP => db:abort_if_pending_migrations => environment
(See full trace by running task with --trace)
```

Very strange one. Investigating.

**Comment by lzap** [comment]
Ok I am able to confirm:
- secret token generates fine
- encryption key ditto
- CR passwords are being encrypted

Upgrade does work fine as well.

:+1:

**Comment by lzap** [comment]
Merged as fab676e for Foreman 1.5. Thanks @domcleal !

**Pull request closed without merging**